ttg/ttg/util/finally.h

```diff
@@ -0,0 +1,57 @@
+#ifndef TTG_UTIL_FINALLY_H
+#define TTG_UTIL_FINALLY_H
+
+///////////////////////////////////////////////////////////////////////////////
+//
+// Copyright (c) 2015 Microsoft Corporation. All rights reserved.
+//
+// This code is licensed under the MIT License (MIT).
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+///////////////////////////////////////////////////////////////////////////////
+
+namespace ttg::detail {
+
+    /* finally, taken from GSL: https://github.com/microsoft/GSL */
+
+    // final_action allows you to ensure something gets run at the end of a scope
+    template <class F>
+    class final_action
```

Comment:
we might want to consider something more fully featured like https://github.com/justusc/deferral

Reply:
at least let's rename to match what will eventually land in C++: https://en.cppreference.com/w/cpp/experimental/scope_exit/scope_exit

Reply:
I took the implementation from https://github.com/tandasat/ScopedResource/ and slightly adopted it.